File: swlog.lua
-- this script has side-effects, so it requires replicate commands mode (deprecated as of Redis v7.0)
redis.replicate_commands()

-- key is a key to associate with this rate limiter
-- limit_events is the maximum number of events to be allowed in a limiting interval. It must be > 0
-- limit_interval is the duration of the limiting interval in milliseconds
-- now is the current Unix time in milliseconds
-- key_ttl is an optional timeout in milliseconds for the key. It must be >= than limit_interval
-- returns limited, remaining, delay
local key = KEYS[1]
local limit_events = tonumber(ARGV[1])
local limit_interval = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local key_ttl = tonumber(ARGV[4])

local trim_time = now - limit_interval
redis.call('ZREMRANGEBYSCORE', key, '-inf', trim_time)

local limited = 1
local count = redis.call('ZCARD', key)
if count < limit_events then
    redis.call('ZADD', key, now, now)
    redis.call('PEXPIRE', key, key_ttl)
    limited = 0
end

local min = tonumber(redis.call('ZRANGE', key, -limit_events, -limit_events)[1])
if not min then
    return { limited, limit_events - count - 1, 0 }
end
return { limited, 0, min - trim_time }

File: swlog.go
package throttle

import (
	"context"
	"crypto/sha1" //nolint:gosec
	_ "embed"     // embed lua script
	"encoding/hex"
	"io"
	"sync"
	"time"
)

//go:embed swlog.lua
var swLogScript string

// SWLogLimiter implements a rate limiter using the Sliding-Window Log algorithm.
// It works with a 1 ms resolution.
type SWLogLimiter struct {
	rds    Rediser
	script script
	keyTTL time.Duration
	clock  func() time.Time

	mu  sync.Mutex
	lim Limit
}

// NewSWLogLimiter returns a new configured [SWLogLimiter].
func NewSWLogLimiter(rds Rediser, limit Limit, opts ...Option) (*SWLogLimiter, error) {
	if err := limit.Valid(); err != nil {
		return nil, err
	}

	options := options{
		keyTTL: 1 * time.Second,
		clock:  time.Now,
	}
	for _, o := range opts {
		o.apply(&options)
	}

	h := sha1.New() //nolint:gosec
	_, _ = io.WriteString(h, swLogScript)

	script := script{
		rds:    rds,
		script: swLogScript,
		sha:    hex.EncodeToString(h.Sum(nil)),
	}

	return &SWLogLimiter{
		rds:    rds,
		script: script,
		keyTTL: options.keyTTL,
		clock:  options.clock,
		lim:    limit,
		mu:     sync.Mutex{},
	}, nil
}

// Allow determines whether the event for the specified key is permitted at the current time.
//
//nolint:forcetypeassert
func (l *SWLogLimiter) Allow(ctx context.Context, key string) (Status, error) {
	l.mu.Lock()
	lim := l.lim
	now := l.clock()
	ttl := max(lim.Interval+l.keyTTL, lim.Interval)
	l.mu.Unlock()

	if lim.Events == 0 {
		return Status{Limited: true, Remaining: 0, Delay: Inf}, nil
	}

	keys := []string{key}
	args := []any{lim.Events, lim.Interval.Milliseconds(), now.UTC().UnixMilli(), ttl.Milliseconds()}
	v, err := l.script.exec(ctx, keys, args...)
	if err != nil {
		return Status{}, err
	}
	values := v.([]any)
	return Status{
		Limited:   values[0].(int64) != 0,
		Remaining: int(values[1].(int64)),
		Delay:     time.Duration(values[2].(int64)) * time.Millisecond, //nolint:durationcheck
	}, nil
}

// Limit returns the current limit.
func (l *SWLogLimiter) Limit() Limit {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.lim
}

// SetLimit sets a new limit.
func (l *SWLogLimiter) SetLimit(_ context.Context, newLimit Limit) error {
	if err := newLimit.Valid(); err != nil {
		return err
	}

	l.mu.Lock()
	defer l.mu.Unlock()
	l.lim = newLimit
	return nil
}

// Reset clears all limitations and previous usage for the specified keys.
// If no keys are provided, it's a no-op.
func (l *SWLogLimiter) Reset(ctx context.Context, keys ...string) error {
	if len(keys) == 0 {
		return nil
	}
	_, err := l.rds.Del(ctx, keys...)
	return err
}
